src/coord/src/coord.rs

```diff
@@ -998,10 +999,47 @@ impl Coordinator {
         }
     }
 
+    fn validate_update_iter(
+        upper: &mut MutableAntichain<Timestamp>,
+        changes: ChangeBatch<Timestamp>,
+        num_workers: usize,
+    ) -> Vec<(Timestamp, i64)> {
+        let old_frontier = upper.frontier().to_owned();
+        let mut total_count = 0;
+
+        let frontier_changes = upper.update_iter(changes.clone().drain()).collect();
+
+        // Make sure no times in `upper` have a negative multiplicity
+        for (t, _) in changes.into_inner() {
+            let count = upper.count_for(&t);
+            assert!(count >= 0);
+            assert!(count as usize <= num_workers);
+            total_count += count;
+        }
+
+        // TODO: this doesn't really check what we want it to check, in that
+        // you could have many changes that increment many distinct times in distinct
+        // ChangeBatches, so total_count stays at 1, but the actual sum of all multiplicities
+        // for times contained in the antichain is >> 1.
+        assert!(total_count as usize <= num_workers);
+
+        // TODO: get rid of this duplicate operation
+        let new_frontier = upper.frontier().to_owned();
+        assert!(<_ as PartialOrder>::less_equal(
+            &old_frontier,
+            &new_frontier
+        ));
+
+        frontier_changes
+    }
+
     /// Updates the upper frontier of a named view.
-    fn update_upper(&mut self, name: &GlobalId, mut changes: ChangeBatch<Timestamp>) {
+    fn update_upper(&mut self, name: &GlobalId, changes: ChangeBatch<Timestamp>) {
+        let num_workers = self.num_workers();
         if let Some(index_state) = self.indexes.get_mut(name) {
-            let changes: Vec<_> = index_state.upper.update_iter(changes.drain()).collect();
+            let changes: Vec<_> =
```

Comment:
You can remove the type helper here since it was only needed by collect, which is now gone.